packages/calypso-e2e/src/lib/utils/validate-translations.ts

```diff
@@ -0,0 +1,121 @@
+	translations: {
+		translation_0: string;
+	}[];
+};
+
+/**
+ * Fetch translations for originals.
+ */
+async function fetchTranslations(
+	page: Page,
+	originals: OriginalString[],
+	locale: string
+): Promise< Translation[] | null > {
+	const params = {
+		url: GLOTPRESS_ORIGINALS_ENDPOINT,
+		data: {
+			project: GLOTPRESS_PROJECT,
+			locale_slug: locale,
+			original_strings: JSON.stringify( originals ),
+		},
+	};
+
+	return page.evaluate( async ( { url, data } ) => {
+		const body = new URLSearchParams();
+		for ( const key in data ) {
+			body.append( key, data[ key as 'project' | 'locale_slug' | 'original_strings' ] );
+		}
+
+		return window
+			.fetch( url, {
+				method: 'POST',
+				headers: {
+					'content-type': 'application/x-www-form-urlencoded',
+				},
+				body: body.toString(),
+			} )
+			.then( ( res ) => res.json() )
+			.then( ( res ) => {
+				delete res.originals_not_found;
+				return res;
+			} )
+			.catch( () => null );
+	}, params );
+}
+
+/**
+ * Get elements translations.
+ */
+async function getElementsTranslations(
+	page: Page,
+	elements: ElementHandle[],
+	locale: string
+): Promise< Translation[] | null > {
+	// Default locale doesn't have translations
+	if ( locale === 'en' ) {
+		return null;
+	}
+
+	const originals: OriginalString[] = await Promise.all(
+		elements.map( async ( element ) => {
+			const singular = ( await element.getAttribute( 'data-e2e-string' ) ) || '';
+			return { singular };
+		} )
+	);
+
+	return fetchTranslations( page, originals, locale );
+}
+
+/**
+ * Validate translations on current page.
+ *
+ * @param {Page} page The underlying page
+ */
+export async function validatePageTranslations( page: Page, locale: string ): Promise< void > {
+	const translatableElements = await page.$$( selectors.translatableElement );
+	const translations = await getElementsTranslations( page, translatableElements, locale );
+
+	for ( const element of translatableElements ) {
+		const singular = await element.getAttribute( 'data-e2e-string' );
+		// In order to test translations with placeholders, parameters must be passed as JSON encoded string.
+		// For example, `<div data-e2e-string="'Hello, %s!'" data-e2e-string-params={[ 'Jane' ]}}>{ sprintf( __( 'Hello, %s!' ), 'Jane' ) }</div>`.
+		const params = await element.getAttribute( 'data-e2e-string-params' );
+
+		let translation = singular;
+
+		// Translation for default locale should match the original.
+		if ( locale !== 'en' ) {
+			translation =
+				translations?.find( ( entry: Translation ) => entry.original.singular === singular )
+					?.translations?.[ 0 ]?.translation_0 || null;
+		}
+
+		if ( ! translation ) {
+			translation = '';
+		}
+
+		if ( params ) {
+			translation = sprintf( translation, ...JSON.parse( params ) );
+		}
+
+		await page.waitForFunction(
+			( { element, translation } ) =>
+				( element as HTMLElement )?.innerText?.trim()?.includes( translation?.trim() || '' ),
+			{ element, translation }
+		);
```

Comment:
Is there any reason for using `waitForFunction` here? I don't think it should be necessary, and to me it slightly muddles the intent of what you're doing in the callback, and makes any failing errors less clear and debuggable. I think these lines could be clearer as just a set of more explicit steps that throws a more detailed error if something doesn't match or isn't right:

```typescript

const actualText = ( element as HTMLElement ).innerText;

 if ( ! actualText  ) {
    throw new Error(`Translatable element had no inner text. HTML of element: ${ element.outerHTML }`;
}

if ( ! ( element as HTMLElement ).innerText.trim().includes( translation.trim() ) {
     throw new Error( `Translations did not match! Expected: ${ translation } --- Actual: ${ actualText }` );
}

```

Reply:
I'm using the `waitForFunction` method, because, in theory, changing the language is an asynchronous action that includes fetching the language, loading it into the translation library and re-rendering the components that use translate calls. So in some cases, it's possible that reading the `innerText` of an element immediately after changing the language would return the English string if the element hasn't re-rendered yet.
Errors are good addition, added them in https://github.com/Automattic/wp-calypso/pull/57707/commits/4672e76befb6a1f561b7b8ca8ba7ff8849101301#diff-c9f4039cb8be26088b4672860ea55820f2f13bbc08555f87923c9b181c2a43dbR123-R135.

Reply:
Oh sure, that makes sense! Good thinking, and thanks for the explanation! 👍